### DISCUSSION TITLE
`atomFamily` that auto cleans up atoms no longer in use

### DISCUSSION THREAD

**Discussion by jvliwanag** [pmndrs/jotai#2775] [Ideas] [answered]
Hi,

I'm trying out this formulation of `atomFamily` that stores atoms on a Map of weak references. Clean-up is done by registration onto a `FinalizationRegistry`.

```typescript
export function atomFamily<K, V extends object>(
  create: (key: K) => V,
): (key: K) => V {
  let cache = new Map<K, WeakRef<V>>();

  let registry = new FinalizationRegistry<K>((key) => {
    cache.delete(key);
  });

  return (key) => {
    let value = cache.get(key)?.deref();
    if (value === undefined) {
      value = create(key);

      registry.register(value, key);

      let weakRef = new WeakRef(value);
      cache.set(key, weakRef);
    }

    return value;
  };
}
```

What do you guys think? Any flaws on this approach?

**Comment by dmaskasky** [comment, marked as answer]
This would totally work and I actually wrote this code before.

But it has a huge problem. What if an atom gets dereferenced too early? Then the next time you call `fooFamily(3)`, you will get a new atom and the previous state would be lost.

**Reply by jvliwanag** [reply]
Ah. :D makes sense only for derived atoms I guess, in which case, it isn't really an `atomFamily`.

Thanks for the guidance!
